# Hypothesis testing problems

## Exercise 1

You are a nutritionist researching two different types of diets to see if there is a significant difference in weight loss after one month. You choose two random groups of people; one group follows the first diet, and the other follows the second. At the end of the month, you record the weight loss (in kg) for each person:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

With these data, it seeks to answer the following question: Is there a significant difference in average weight loss between people who followed the first diet and those who followed the second diet?

To draw conclusions, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the test to test the hypothesis. You can use a Student's t-test.
- Analyze the conclusions.

In [35]:
import numpy as np
import matplotlib.pyplot as plt
from math import *
import scipy.stats as ss
from scipy.stats import t

d1=np.array([2,2.5,3,2.8,2.3,2.7,2.5])
d2=np.array([3,3.2,3.1,2.9,2.8,3,3.2])

alfa=0.05
conf=1-alfa
n1=len(d1)
n2=len(d2)
df=n1+n2-2  # Grados de Libertad n1+n2-2
#print(df)
m1=np.mean(d1)
m2=np.mean(d2)
m1,m2
v1=np.var(d1,ddof=1) # En el calculo de la varianza de una muestra se usa un grado de libertad, es decir, se divide entre n-1 en la formula
v2=np.var(d2,ddof=1)
valor_critico = t.ppf(q=0.95 + 0.05/2, df=df, loc=0, scale=1)
print('Valor Critico (t con alfa=0.05, grados de libertad 12, dos colas): ','-',round(valor_critico,4),'y +',round(valor_critico,4))
var_conj=((n1-1)*v1+(n2-1)*v2)/df
tprueba=(m1-m2)/(sqrt(var_conj*(1/n1+1/n2)))
print('Estadistico de prueba t: ',round(tprueba,4))
valor_2=t.cdf(tprueba,12)
print('Probabilidad Acumulada en el Estadistico de Prueba t: ', round(tprueba,4),'Probabilidad (Incluyendo dos colas): ',round(valor_2*2,6))



Valor Critico (t con alfa=0.05, grados de libertad 12, dos colas):  - 2.1788 y + 2.1788
Estadistico de prueba t:  -3.5383
Probabilidad Acumulada en el Estadistico de Prueba t:  -3.5383 Probabilidad (Incluyendo dos colas):  0.004083


## Conclusiones

Vamos a suponer que los datos representan a dos muestras homogeneas, independientes y de tamano menor a 30, por lo cual, se puede usar para la prueba de hipotesis (diferencia de dos medias) la distribucion t. Adicionalmente, se parte del hecho que la poblacion obedece a una Dist. Normal. En este caso se formulara la siguiente hipotesis nula:

Ho: Media Diet 1 = Media Diet 2, contra la siguiente hipotesis alternativa
Ha: Media Diet 1 es diferente a Media Diet 2

Se asume un nivel de significancia de 0.05 (nivel de confianza de 95%). En este caso es una prueba de dos colas. Se debe calcular el valor critico para aceptar o rechazar la hipotesis nula, en este caso el valor es +/-2.1788.

El estadistico de prueba t es igual a -3.5383, el cual es menor a -2.1788, la probabilidad acumulada en el estadistico de prueba t es         0.004083 mucho menor al nivel de significancia de 0.05, con lo cual, se tiene bastante evidencia para rechazar la hipotesis nula y aceptar la hipotesis alternativa, es decir, que el hacer una dieta u otra puede incidir en la mayor perdida de peso. 

Viendo la solucion propuesta, encuentro que la biblioteca de Scipy en su modulo de stats tiene una funcion para comparar la igual de medias en dos muestrasfd, arrojando los valores del estadistico de prueba y la prob acumulada asociada, es el mejor camino, obteniendo las mismas conclusiones, excelente....

t_value, p_value = stats.ttest_ind(d1, d2)

print(f"t-value: {t_value}")
print(f"p-value: {p_value}")


## ANOVA

**ANOVA** (*Analysis of Variance*) is a statistical technique used to compare the measures of two or more groups. The idea behind ANOVA is to decompose the total variability in the data into two components: between-group variability and within-group variability:

- **Between-group variability**: This variability refers to the differences between the group means. If this variability is considerably larger than the within-group variability, it could be an indication that at least one of the group means is different.
- **Within-group variability**: This variability refers to the dispersion of the data within each group. If all groups have similar variability, then any noticeable difference in group means could be considered significant.

Hypotheses in ANOVA typically include:

- **Null hypothesis** ($H₀$): The means of all groups are equal.
- **Alternative hypothesis** ($H₁$): At least one of the group means is different.

If the ANOVA test result is significant (e.g., a p-value less than a threshold such as 0.05), this suggests that at least one group mean is different.

## Exercise 2

A farmer decides to test three different types of fertilizers to determine if one is superior in terms of corn production. The farmer plants corn on 15 identical plots and uses all three fertilizers (5 plots for each type). At the end of the season, he measures the corn yield (in kg) of each plot, with the following result:

| Fertilizer 1 | Fertilizer 2 | Fertilizer 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

With this data, he seeks to answer the following question: Is there a significant difference in average corn yield between the three types of fertilizers?

To help you, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the ANOVA test.
- Analyze the conclusions.
- If one fertilizer is better than another, how can we know it?

In [131]:
g1=([20,21,20,19,20])
g2=([22,21,23,22,21])
g3=([24,23,22,23,24])
g=np.array([g1,g2,g3])
gt=np.transpose(g)
fil, col=gt.shape
alf=0.05
q=1-alf
dfn=col-1
dfd=fil*col-col
f_val, p_val = ss.f_oneway(gt[:,0],gt[:,1],gt[:,2])
f_val, p_val
valorCriticpF=ss.f.ppf(q=q, dfn=dfn, dfd=dfd)
print('Valor Critico del Estadistico F: ',round(valorCriticpF,3))
print('Valor Estadistico de Prueba F: ',round(f_val,3))
print('Valor Probabilidad a la derecha del Estadistico de Prueba F: ',round(p_val,5))

#q: 1 menos el nivel de significancia a utilizar
#dfn : los grados de libertad del numerador
#dfd : los grados de libertad del denominador

Valor Critico del Estadistico F:  3.885
Valor Estadistico de Prueba F:  20.316
Valor Probabilidad a la derecha del Estadistico de Prueba F:  0.00014


In [143]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import pandas as pd
data=g1+g2+g3
fert_df= pd.DataFrame({'Valores': data,
                   'Tipo_Fert': np.repeat(['F1', 'F2', 'F3'], repeats=5)}) 
fert_df
tukey = pairwise_tukeyhsd(endog=fert_df['Valores'], groups=fert_df['Tipo_Fert'], alpha=0.05)
print(tukey)
print('La produccion media en Kg usando el fertilizante 1 fue de: ',np.mean(g1))
print('La produccion media en Kg usando el fertilizante 2 fue de: ',np.mean(g2))
print('La produccion media en Kg usando el fertilizante 3 fue de: ',np.mean(g3))

Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj  lower  upper  reject
--------------------------------------------------
    F1     F2      1.8 0.0099 0.4572 3.1428   True
    F1     F3      3.2 0.0001 1.8572 4.5428   True
    F2     F3      1.4 0.0409 0.0572 2.7428   True
--------------------------------------------------
La produccion media en Kg usando el fertilizante 1 fue de:  20.0
La produccion media en Kg usando el fertilizante 2 fue de:  21.8
La produccion media en Kg usando el fertilizante 3 fue de:  23.2


## Conclusiones

En primer lugar se establece como hiportesis nula que las tres medias de los gurpos de la muestra son iguales contra una hipotesis alternativa de que al menos, una de las medias es diferente. Vamos a suponer que las muestras son independientes, obedecen a distribuciones normales y cumplen con la propiedad de homocedasticidad (el error cometido por el modelo tiene siempre la misma varianza).

Se puede observar que el valor del estadistico de prueba F, calculado con la muestra es mucho mayor que el valor critico de F dado un nivel de significancia del 0.05, por lo cual, se rechaza la hipotesis nula y se acepta la alternativa, incluso la probabilidad a la derecha del estadistico de prueba F es 0.00014, el cual es mucho menor que el nivel de significancia (0,05) asumido en el modelo.

En relacion a la comparacion de las medias entre grupos, se corrio la prueba de tukey para determinar si habia diferencias en la produccion promedio de maiz al usar fertilizantes distintos (se comparo el F1 versus F2 y F3, F2 versus F3), el test arrojo que existen diferencias en todos los cruces (el valor de p-adj dio menor que alfa=0.05), por lo cual, si vemos la produccion promedio en la data, se observa que usar el fertilizante 3, es la mejor opcion, puesto que su media es la mas alta con respecto a F1 y F2, esta conclusion de las diferencias de las medias puede verse en la tabla de tukey en la columna meandiff, por ejemplo la diferencia de la media en F1 y F2 es igual a F2 - F1 = 21.8-20 = 1.8, la media de F2 es mayor que F1.